v4/weaviate/connection/rest.go

```diff
@@ -27,12 +29,43 @@ func NewConnection(scheme string, host string, httpClient *http.Client, headers
 	if client == nil {
 		client = &http.Client{}
 	}
-
-	return &Connection{
+	connection := &Connection{
 		basePath:   scheme + "://" + host + "/" + apiVersion,
 		httpClient: client,
 		headers:    headers,
 	}
+
+	// background goroutine that periodically refreshes the auth token.
+	// The oauth2 package only refreshes the Tokens on new http requests => if there is no request for the lifetime of
+	// the refresh token the client will become de-authenticated without this.
+	go func(con *Connection) {
+		transport, ok := con.httpClient.Transport.(*oauth2.Transport)
+		if !ok {
+			return
+		}
+		for {
+			token, err := transport.Source.Token()
+			if err != nil {
+				log.Printf("Error during token refresh, getting token: %v", err)
+				return
+			}
+			// there is no point in manual refreshing if there is no refresh token
+			if token.RefreshToken == "" {
+				log.Println("No refresh token, exiting background goroutine")
+				return
+			}
+
+			timeToSleep := token.Expiry.Sub(time.Now()) - time.Second/10
+			time.Sleep(timeToSleep)
```

Comment:
I would suggest to refactor this goroutine to use channels with ticker instead of sleep. it would make also a code a little bit more readable

Reply:
refactored the goroutine a bit, please have another look